### PR TITLE
Move axis obj

### PR DESCRIPTION
**Pull request by uramirez8707**
**Description**
- Moves the axis and subaxis objects in the controller
- Setups the public diag_axis interfaces in the controller. So far only the interfaces that are used in the one SPEAR and AM4 experiment that I ran have been updated ...
- Replaces `edges` (the edge axis id) with `edges_name`. The edges axis id was only used to get the name of the edges axis so it can be added as a attribute. 

Note that this forces `diag_axis_mod` to depend on `fms_diag_object_mod`. Since `diag_utils_mod` depends on `diag_axis_mod`, `diag_utils_mod` cannot be used anywhere in the modern_diag mods. 

Fixes # (issue)

**How Has This Been Tested?**
CI
0 day run of SPEAR

**Checklist:**
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published in downstream modules
- [x] New check tests, if applicable, are included
- [x] `make distcheck` passes

### CODE REVIEW

**Review by thomas-robinson — CHANGES_REQUESTED** [review]

File: diag_manager/fms_diag_field_object.F90 [hunk 1]
```diff
@@ -210,7 +206,7 @@ subroutine fms_register_diag_field_obj &
 !> Add axis and domain information
   if (present(axes)) then
```
[the remark this answers is not in the document]
Comment (uramirez8707) [reply diag_manager/fms_diag_field_object.F90:207]: diag_axis is not optional, the array of diag_axis will always be passed into the registered. If the variable is a scalar, it just won't use it.

File: diag_manager/fms_diag_file_object.F90 [hunk 2]
```diff
@@ -27,10 +27,10 @@ module fms_diag_file_object_mod
 #ifdef use_yaml
 use fms2_io_mod, only: FmsNetcdfFile_t, FmsNetcdfUnstructuredDomainFile_t, FmsNetcdfDomainFile_t
 use diag_data_mod, only: DIAG_NULL, NO_DOMAIN, max_axes, SUB_REGIONAL, get_base_time, DIAG_NOT_REGISTERED
-use diag_util_mod, only: diag_time_inc
+!TODO cross dependency use diag_util_mod, only: diag_time_inc
```
[the remark this answers is not in the document]
Comment (uramirez8707) [reply diag_manager/fms_diag_file_object.F90:30]: oh I meant circular dependency?

fms_diag_file_object_mod depends on diag_utils_mod

but diag_utils_mod depends on diag_axis_mod which depends on fms_diag_object_mod which depends on fms_diag_file_object_mod 

so right now they depend on each other. 


We will need to move any subroutine that diag_utils_mod needs (like diag_time_inc to another module or something).

File: diag_manager/fms_diag_axis_object.F90 [hunk 3]
```diff
@@ -365,22 +377,19 @@ function set_subaxis(obj, bounds) &
...
     enddo
 
-    !< TO DO: everything
+    !TODO: everything
     obj%nsubaxis = obj%nsubaxis + 1
-
-    nsubaxis_objs = nsubaxis_objs + 1
-    sub_axes_id = nsubaxis_objs
-    !< TO DO: set the parent_axis_id
+    sub_axes_id = -999
```
[the remark this answers is not in the document]
Comment (uramirez8707) [reply diag_manager/fms_diag_axis_object.F90:382]: yeah that is part of the !TODO: everything

File: diag_manager/fms_diag_object.F90 [hunk 4]
```diff
@@ -39,7 +41,11 @@ module fms_diag_object_mod
 !TODO: Remove FMS prefix from variables in this type
   class(fmsDiagFileContainer_type), allocatable :: FMS_diag_files (:) !< array of diag files
   class(fmsDiagField_type), allocatable :: FMS_diag_fields(:) !< Array of diag fields
+  class(fmsDiagAxis_type), allocatable :: diag_axis(:) !< Array of diag_axis
+  class(fmsDiagSubAxis_type), allocatable :: diag_sub_axis(:) !< Array of diag_sub_axis
   integer, private :: registered_variables !< Number of registered variables
+  integer, private :: registered_axis !< Number of registered axis
+  integer, private :: registered_subaxis !< Number of registered subAxis
```
[the remark this answers is not in the document]
Comment (uramirez8707) [reply diag_manager/fms_diag_object.F90:48]: I don't think that is necessary. 

The subaxis needs information about a parent axis (i.e the units, cart_name, data, etc ...) so if we do
```
type, extends(fmsDiagAxis_type) :: fmsDiagSubAxis_type
```
We will have to copy all the info from the parent axis to the sub axis object.
Rather than doing that I just set up a parent_axis_id variable in the sub axis object that we can just to get all the information we need. 

I suppose we have something like:

```F90
type :: fmsDiagAxisContainer_type                                                                                       
  class(fmsDiagAxis_type), allocatable :: axis                                                                          

! PROCEDURES: diag_parent_axis_init
! PROCEDURES: diag_sub_axis_init
end type 

type :: fmsDiagAxis_type                                                                                                
  integer :: axis_id                                                                                                    
! PROCEDURES :: get_axis_length, write_axis_data, write_axis_metadata                                                   
end type                                                                                                                
                                                                                                                        
type, extends(fmsDiagAxis_type) :: fmsDiagParentAxis_type                                                               
  integer :: cart !< axis_data, axis_name, etc etc                                                                      
end type                                                                                                                
                                                                                                                        
type, extends(fmsDiagAxis_type) :: fmsDiagSubAxis_type                                                                  
  integer :: parent_axis_id                                                                                             
  !< other subaxis stuff                                                                                                
end type 
```

The diag_axis_init will allocate `axis` to be fmsDiagParentAxis_type and add all the info inside of it
Whenever we need a subaxis we can allocate `axis` to be fmsDiagSubAxis_type and fill it in. 

What do you think?

File: diag_manager/docs_uml/classDiagramDiagObjects.drawio [hunk 5]
```diff
@@ -7,7 +7,7 @@
...
+        <mxCell id="K0b4yNtIqu9R4qsapTuJ-2" value="baseDate&#xa;character :: experiment_name&#xa;type(fmsDiagFiles_type) :: diag_files (:)&#xa;type(fmsDiagSubregionFiles_type) :: diag_subregion_files (:)&#xa;type(fmsDiagDiurnalFiles_type) :: diag_diurnal_files (:)&#xa;type(fmsDIagVars_type) :: diag_vars (:)&#xa;type(fmsDiagBuffer_type) :: diag_buffers(:) &#xa;type(fmsfmsDiagAxis_typeype) :: diag_axis(:)" style="part=1;html=1;resizeHeight=0;strokeColor=none;fillColor=none;align=left;verticalAlign=middle;spacingLeft=4;spacingRight=4;overflow=hidden;rotatable=0;points=[[0,0.5],[1,0.5]];portConstraint=eastwest;swimlaneFillColor=#ffffff;strokeOpacity=100;whiteSpace=wrap;fontSize=16.7;fontColor=#3a414a;" vertex="1" parent="K0b4yNtIqu9R4qsapTuJ-1">
```
Comment (ngs333) [inline diag_manager/docs_uml/classDiagramDiagObjects.drawio:10]: fmsfms

Comment (thomas-robinson) [inline diag_manager/docs_uml/classDiagramDiagObjects.drawio:10]: Why is this changed?

File: diag_manager/fms_diag_file_object.F90 (hunk 2 above)
[the remark this answers is not in the document]
Comment (thomas-robinson) [reply diag_manager/fms_diag_file_object.F90:30]: I think it needs to be the other way around.  diag_utils should be a low-level routine, and the objects should be above it.  So diag_utils should not depend on fms_diag_axis_mod.  Is there any reason why diag_utils needs the axis information?

File: diag_manager/fms_diag_object.F90 (hunk 4 above)
[the remark this answers is not in the document]
Comment (thomas-robinson) [reply diag_manager/fms_diag_object.F90:48]: I think this is a great idea, but the naming is wrong.  `fmsDiagParentAxis_type` is not a parent, it's a child.  Maybe something like:
```fortran
type :: fmsDiagAxisContainer_type                                                                                       
  class(fmsDiagAxis_type), allocatable :: axis                                                                          

! PROCEDURES: diag_parent_axis_init
! PROCEDURES: diag_sub_axis_init
end type 

type :: fmsDiagAxis_type                                                                                                
  integer :: axis_id                                                                                                    
! PROCEDURES :: get_axis_length, write_axis_data, write_axis_metadata                                                   
end type                                                                                                                
                                                                                                                        
type, extends(fmsDiagAxis_type) :: fmsDiagFullAxis_type                                                               
  integer :: cart !< axis_data, axis_name, etc etc                                                                      
end type                                                                                                                
                                                                                                                        
type, extends(fmsDiagAxis_type) :: fmsDiagSubAxis_type                                                                  
  integer :: parent_axis_id                                                                                             
  !< other subaxis stuff                                                                                                
end type 
```
Now it differentiates between a `full` axis and a `sub` axis. Don't feel like you have to go with this, it's just a suggestion.  But you should definitely not name that type `fmsDiagParentAxis_type`

**Comment by thomas-robinson** [comment]
@uramirez8707 You need to update *THIS* (see what I did there) to resolve the merge conflicts.

**Review by ngs333 — COMMENTED** [review]
@uramirez8707 
You wrote that ```diag_utils_mod``` cannot be used anywhere in the modern_diag mods. But even the modern ```send_data_3d``` needs to use some functions that are in diag_utils_mod. E.g. check_out_of_bounds, check_bounds_are_exact_dynamic. Perhaps such functions need a rewrite for modern_diag? Do we need use of ```#ifdef modern_diag``` for this? Perhaps there is another way?

File: diag_manager/fms_diag_file_object.F90 (hunk 2 above)
[the remark this answers is not in the document]
Comment (uramirez8707) [reply diag_manager/fms_diag_file_object.F90:30]: diag_utils needs diag_axis_mod (the old one) 
https://github.com/NOAA-GFDL/FMS/blob/e8870d4fdcb97388119fa97a2c93a9cdfcc86fc3/diag_manager/diag_util.F90#L54-L57

The diag_axis_init routine is in diag_axis_mod, so diag_axis_mod has to depend on the fms_diag_object so we can do:
https://github.com/NOAA-GFDL/FMS/blob/682281d84c368130030beaaed19c71c4e1fa9bce/diag_manager/diag_axis.F90#L140-L144

**Comment by uramirez8707** [comment]
> @uramirez8707 You wrote that `diag_utils_mod` cannot be used anywhere in the modern_diag mods. But even the modern `send_data_3d` needs to use some functions that are in diag_utils_mod. E.g. check_out_of_bounds, check_bounds_are_exact_dynamic. Perhaps such functions need a rewrite for modern_diag? Do we need use of `#ifdef modern_diag` for this? Perhaps there is another way?

I think we are going to have to move and potentially rewrite any of the thing in diag_utils_mod that we need to another mod. This will also helps us later on when we have to delete all the old stuff.

File: diag_manager/fms_diag_axis_object.F90 [hunk 6]
```diff
@@ -390,7 +399,7 @@ function check_if_subaxis_exists(obj,bounds) &
   !> @return Length of the 2D domain
   function get_length(obj, cart_axis, domain_position, global_length) &
```
[the remark this answers is not in the document]
Comment (uramirez8707) [reply diag_manager/fms_diag_axis_object.F90:400]: This is not a pure function because `mpp_get_compute_domain` which is used is not.

```
error #7137: Any procedure referenced in a PURE procedure, including one referenced via a defined operation or assignment, must have an explicit interface and be declared PURE.   [MPP_DOMAINS_MOD^MPP_GET_COMPUTE_DOMAIN2D]
      if (trim(cart_axis) == "X") call mpp_get_compute_domain(this%Domain2, xsize=length, position=domain_position)

```

File: diag_manager/docs_uml/classDiagramDiagObjects.drawio [hunk 7]
```diff
@@ -7,7 +7,7 @@
...
           <mxGeometry x="1165" y="30" width="615" height="488.25" as="geometry" />
         </mxCell>
-        <mxCell id="K0b4yNtIqu9R4qsapTuJ-2" value="baseDate&#xa;character :: experiment_name&#xa;type(fmsDiagFiles_type) :: diag_files (:)&#xa;type(fmsDiagSubregionFiles_type) :: diag_subregion_files (:)&#xa;type(fmsDiagDiurnalFiles_type) :: diag_diurnal_files (:)&#xa;type(fmsDIagVars_type) :: diag_vars (:)&#xa;type(fmsDiagBuffer_type) :: diag_buffers(:) &#xa;type(fmsDiagAxis_type) :: diag_axis(:)" style="part=1;html=1;resizeHeight=0;strokeColor=none;fillColor=none;align=left;verticalAlign=middle;spacingLeft=4;spacingRight=4;overflow=hidden;rotatable=0;points=[[0,0.5],[1,0.5]];portConstraint=eastwest;swimlaneFillColor=#ffffff;strokeOpacity=100;whiteSpace=wrap;fontSize=16.7;fontColor=#3a414a;" vertex="1" parent="K0b4yNtIqu9R4qsapTuJ-1">
```
[the remark this answers is not in the document]
Comment (uramirez8707) [reply diag_manager/docs_uml/classDiagramDiagObjects.drawio:10]: reverting

File: diag_manager/docs_uml/classDiagramDiagObjects.drawio (hunk 5 above)
Comment (uramirez8707) [reply diag_manager/docs_uml/classDiagramDiagObjects.drawio:10]: reverting

Comment (uramirez8707) [reply diag_manager/docs_uml/classDiagramDiagObjects.drawio:10]: reverting

File: diag_manager/fms_diag_axis_object.F90 [hunk 8]
```diff
@@ -426,70 +435,32 @@ subroutine set_axis_domain(obj, Domain, Domain2, DomainU)
...
+  !< @brief Deallocates the axis and subaxis objects
+  !! @return true if there are axis and subaxis allocated
+  logical function fms_diag_axis_object_end(axis_array, subaxis_array)
+    class(fmsDiagAxis_type)   , allocatable, intent(inout) :: axis_array(:) !< Array of diag_axis
+    class(fmsDiagSubAxis_type), allocatable, intent(inout) :: subaxis_array(:) !< Array of diag_subaxis
 
-    if (axis_id < 0 .and. axis_id > number_of_axis) &
-      call mpp_error(FATAL, "diag_axis_add_attribute: The axis_id is not valid")
+    deallocate(axis_array)
+    deallocate(subaxis_array)
```
[the remark this answers is not in the document]
Comment (uramirez8707) [reply diag_manager/fms_diag_axis_object.F90]: done in https://github.com/NOAA-GFDL/FMS/pull/1024/commits/f09879e97c7fef3c5900880483d43b293d410e31

File: diag_manager/fms_diag_axis_object.F90 [hunk 9]
```diff
@@ -426,70 +435,32 @@ subroutine set_axis_domain(obj, Domain, Domain2, DomainU)
...
-    number_of_axis = number_of_axis + 1
+  !< @brief Allocates the axis and subaxis objects
+  !! @return true if there are axis and subaxis allocated
+  logical function fms_diag_axis_object_init(axis_array, subaxis_array)
+  class(fmsDiagAxis_type)   , allocatable, intent(inout) :: axis_array(:) !< Array of diag_axis
+  class(fmsDiagSubAxis_type), allocatable, intent(inout) :: subaxis_array(:) !< Array of diag_subaxis
 
-    if (number_of_axis > max_axes) call mpp_error(FATAL, &
-      &"diag_axis_init: max_axes exceeded, increase via diag_manager_nml")
+    allocate(axis_array(max_axes))
```
[the remark this answers is not in the document]
Comment (uramirez8707) [reply diag_manager/fms_diag_axis_object.F90:455]: done in https://github.com/NOAA-GFDL/FMS/pull/1024/commits/f09879e97c7fef3c5900880483d43b293d410e31

File: diag_manager/fms_diag_object.F90 (hunk 4 above)
[the remark this answers is not in the document]
Comment (uramirez8707) [reply diag_manager/fms_diag_object.F90]: done in https://github.com/NOAA-GFDL/FMS/pull/1024/commits/f09879e97c7fef3c5900880483d43b293d410e31

File: diag_manager/fms_diag_axis_object.F90 [hunk 10]
```diff
@@ -365,22 +377,19 @@ function set_subaxis(obj, bounds) &
...
-    nsubaxis_objs = nsubaxis_objs + 1
-    sub_axes_id = nsubaxis_objs
-    !< TO DO: set the parent_axis_id
+    sub_axes_id = -999
   end function
 
   !!!!!!!!!!!!!!!!!! SUB AXIS PROCEDURES !!!!!!!!!!!!!!!!!
   !> @brief Check if a subaxis was already defined
   !> @return Flag indicating if a subaxis is already defined
   function check_if_subaxis_exists(obj,bounds) &
```
[the remark this answers is not in the document]
Comment (uramirez8707) [reply diag_manager/fms_diag_axis_object.F90]: done in https://github.com/NOAA-GFDL/FMS/pull/1024/commits/bd1bbef122fde24f580be487878d97291280b1c8

**Review by ngs333 — APPROVED** [review]

File: diag_manager/diag_axis.F90 [hunk 11]
```diff
@@ -590,8 +589,12 @@ SUBROUTINE get_diag_axis_name(id, name)
     INTEGER         , INTENT(in)  :: id !< Axis ID
     CHARACTER(len=*), INTENT(out) :: name !< Axis short name
```
Comment (thomas-robinson) [inline diag_manager/diag_axis.F90:590]: This variable name needs to be changed

File: diag_manager/fms_diag_file_object.F90 (hunk 2 above)
[the remark this answers is not in the document]
Comment (thomas-robinson) [reply diag_manager/fms_diag_file_object.F90:30]: I've just spent 10 minutes sighing.  Why?  Why?  Why?

OK, mean I *think* I have an idea that I hate, but idk what else to do here.  I think we should make (yet another) module that has no dependencies and houses constants and the low-level routines that are used by other modules.  Then diag_data can just use those values and whatever else from this lower-level module.  

Here comes another meeting to discuss it...

File: diag_manager/fms_diag_file_object.F90 [hunk 12]
```diff
@@ -574,8 +578,9 @@ subroutine add_start_time(this, start_time)
     !! simply update it with the start_time and set up the *_output variables
     this%start_time = start_time
     this%last_output = start_time
-    this%next_output = diag_time_inc(start_time, this%get_file_freq(), this%get_file_frequnit())
-    this%next_next_output = diag_time_inc(this%next_output, this%get_file_freq(), this%get_file_frequnit())
+    !TODO circular dependency
```
Comment (thomas-robinson) [inline diag_manager/fms_diag_file_object.F90:581]: What is this circular dependency?

File: diag_manager/fms_diag_object.F90 [hunk 13]
```diff
@@ -192,6 +206,8 @@ integer function fms_register_diag_field_obj &
   nullify (fileptr)
   nullify (fieldptr)
   deallocate(diag_field_indices)
+#else
```
[the remark this answers is not in the document]
Comment (thomas-robinson) [reply diag_manager/fms_diag_object.F90:208]: We still need to discuss this.  Most of this file should be `#ifdef`ed out.  I think there are too many and it's too complicated.

**Review by thomas-robinson — CHANGES_REQUESTED** [review]
One of the problems I have with this is when the old axis stuff is removed, we will still be stuck with diag_axis.F90.  Then we will have to deal with moving everything again.  If we have `use_modern_diag=.true`, will we be calling `  INTEGER FUNCTION get_axis_length(id)` and the other functions/routines of the old diag manager?  It also creates this dependency issue with diag_data_mod.  I think we need to talk about this a little more.  

There are a few coding things that can be updated too like `Pure` routines.

File: diag_manager/fms_diag_file_object.F90 (hunk 2 above)
[the remark this answers is not in the document]
Comment (uramirez8707) [reply diag_manager/fms_diag_file_object.F90:30]: The dependency is with diag_utils not diag_data, but yeah we make another modules that has utils that are in diag_utils_mod that the new diag_manager needs (like diag_time_inc)

File: diag_manager/fms_diag_axis_object.F90 (hunk 9 above)
[the remark this answers is not in the document]
Comment (uramirez8707) [reply diag_manager/fms_diag_axis_object.F90:455]: but fms_diag_axis_object_init is the function's output. I don't think it passes in a value?

**Comment by uramirez8707** [comment]
> One of the problems I have with this is when the old axis stuff is removed, we will still be stuck with diag_axis.F90. Then we will have to deal with moving everything again. If we have `use_modern_diag=.true`, will we be calling ` INTEGER FUNCTION get_axis_length(id)` and the other functions/routines of the old diag manager? It also creates this dependency issue with diag_util_mod. I think we need to talk about this a little more.

We are going to be stuck with diag_axis.F90 either way because the model components are using it directly. 
https://github.com/NOAA-GFDL/MOM6/blob/66ba6a7ad78a923212c5d317e1c3cc5609961062/config_src/infra/FMS2/MOM_diag_manager_infra.F90#L11-L14

I think we just be removing all the code except for the the code inside the `if (use_modern_diag)`

File: diag_manager/fms_diag_axis_object.F90 (hunk 3 above)
[the remark this answers is not in the document]
Comment (uramirez8707) [reply diag_manager/fms_diag_axis_object.F90:393]: This is going to be done in another PR because it will require a lot of changes.

File: diag_manager/fms_diag_axis_object.F90 [hunk 14]
```diff
@@ -367,30 +390,27 @@ function set_subaxis(this, bounds) &
...
+    logical                                   :: exists
 
     !< TO DO: compare bounds
     exists = .false.
-  end function
+  end function check_if_subaxis_exists
 
   !> @brief Get the length of a 2D domain
   !> @return Length of the 2D domain
   function get_length(this, cart_axis, domain_position, global_length) &
```
[the remark this answers is not in the document]
Comment (uramirez8707) [reply diag_manager/fms_diag_axis_object.F90:411]: This is not pure.. it uses `mpp_get_compute_domain` which is not pure or a function

File: diag_manager/fms_diag_file_object.F90 (hunk 12 above)
Comment (uramirez8707) [reply diag_manager/fms_diag_file_object.F90:581]: It is the same circular dependency as above. `diag_time_inc` is in `diag_utils_mod`

`diag_utils_mod` depends on `diag_axis_mod`

`diag_axis_mod` depends on `fms_diag_object_mod`

and `fms_diag_object_mod` depends on `fms_diag_file_object_mod`

It's a circle. 

We will need to move `diag_time_inc` to a lower level module that only depends on diag_data_mod.

File: diag_manager/fms_diag_axis_object.F90 [hunk 15]
```diff
@@ -426,70 +446,27 @@ subroutine set_axis_domain(this, Domain, Domain2, DomainU)
...
-
-    if (axis_id < 0 .and. axis_id > number_of_axis) &
-      call mpp_error(FATAL, "diag_axis_add_attribute: The axis_id is not valid")
-
-    call axis_obj(axis_id)%add_axis_attribute(att_name, att_value)
-  end subroutine fms_diag_axis_add_attribute
+  end function fms_diag_axis_object_end
 
   !> @brief Check if a cart_name is valid and crashes if it isn't
   subroutine check_if_valid_cart_name(cart_name)
```
[the remark this answers is not in the document]
Comment (uramirez8707) [reply diag_manager/fms_diag_axis_object.F90:472]: cart_name is related to the diag_axis, so I think it should stay in fms_diag_axis_object.

File: diag_manager/fms_diag_axis_object.F90 [hunk 16]
```diff
@@ -337,8 +342,8 @@ end subroutine write_axis_data
   !> @return axis length
   function get_axis_length(this) &
```
[the remark this answers is not in the document]
Comment (uramirez8707) [reply diag_manager/fms_diag_axis_object.F90]:
```
fms_diag_axis_object.F90:352:20:

  352 |       axis_length = this%axis_domain%length(this%cart_name, this%domain_position, this%length)
      |                    1
Error: Reference to impure function 'get_length' at (1) within a PURE procedure

```
this%axis_domain%length is not pure because it is using `mpp_get_compute_domain`

File: diag_manager/fms_diag_axis_object.F90 (hunk 9 above)
[the remark this answers is not in the document]
Comment (thomas-robinson) [reply diag_manager/fms_diag_axis_object.F90:457]: I was thinking "What the heck is he talking about?" and then I realized that I have poorly named the logical module variable here.  
```fortran
logical :: fms_diag_axis_object_initialized = .false. !< module variable initialized to false
logical function fms_diag_axis_object_init(axis_array)
  class(fmsDiagAxisContainer_type)   , allocatable, intent(inout) :: axis_array(:) !< Array of diag_axis￼
!!! Add check if already init !!!
  if (fms_diag_axis_object_initialized) return
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
  number_of_axis = 0		￼    
  if (allocated(axis_array)) call mpp_error(FATAL, "The diag_axis container is already allocated")
  allocate(axis_obj(max_axes))
  allocate(axis_array(max_axes))
  fms_diag_axis_object_init = .true.
  fms_diag_axis_object_initialized = .true.
end function fms_diag_axis_object_init
```

**Review by thomas-robinson — APPROVED** [review]
